File: 2019/01/solution.hs
#!/usr/bin/runghc

main :: IO ()
main = do
    input <- readFile "input.txt"
    putStrLn $ task1 input
    putStrLn $ task2 input

task1 :: String -> String
task1 input =
    show $ sum $ map fuel $ masses input

task2 :: String -> String
task2 input =
    show $ sum $ map fuel_recursive $ masses input

masses :: String -> [Int]
masses input = map read $ lines input

fuel :: Int -> Int
fuel mass = mass `div` 3 - 2

fuel_recursive :: Int -> Int
fuel_recursive mass =
    let f = fuel(mass)
    in if f > 0 then f + fuel_recursive(f) else 0

File: 2019/01/solution.py
#!/usr/bin/python3

def fuel(mass):
    return mass // 3 - 2

def masses(input):
    return (int(line.strip()) for line in input.splitlines() if line)

def task1(input):
    return sum(fuel(mass) for mass in masses(input))

def fuel_recursive(mass):
    f = fuel(mass)
    return (f + fuel_recursive(f)) if f > 0 else 0

def task2(input):
    return sum(fuel_recursive(mass) for mass in masses(input))

input = open('input.txt', 'r').read()
print(task1(input))
print(task2(input))
